docs/dsl-cheatsheet.md

```diff
@@ -0,0 +1,83 @@
+
+        // Handle external "input", called Actions
+        on<Action1>{ action, state -> ... } //  Handle an action
+        on<Action2>(ExecutionPolicy){ action, state -> ... } //  You can have multiple on<Action> blocks. Optinal specify ExecutionPolicy
+       
+        // Do something when you enter the state.
+        onEnter{ state -> ... } // called exactly one time when the given state has been entered
+        onEnter{ state -> ... } // You can have multiple onEnter blocks
+        
+        // Collect a Flow (from coroutine package) as long as the state machine is in the state (see inState<State>)
+        collectWhileInstate(flow1) { valueEmitedFromFlow, state -> ... } // stops flow collection when state is left
+        collectWhileInstate(flow2) { valueEmitedFromFlow, state -> ... } // You can have multiple collectWhileInstate
+
+        // Effects to do something without changing the state (i.e. logging, analytics, ...)
+        onActionEffect<Action1>{ action, state -> ... } // You can have multiple onActionEffect
+        onEnterEffect{ state -> ... } // You can have multiple onEnterEffect
+        collectWhileInStateEffect(flow1){ valueEmitedFromFlow, state -> ... } // You can have multiple collectWhileInstate
+
+        // hierarchical state machines
+        onEnterStartStateMachine(
+            stateMachineFactory = { stateSnapshot : State1  -> OtherStateMachine() },
+            stateMapper = { state : State<State1>, otherStateMachineState : OtherState ->
+                ...
+            }
+        )     
+        onActionStartStateMachine<Action1>(
+            stateMachineFactory = { action, stateSnapshot : State1 -> OtherStateMachine() },
+            stateMapper = { state : State<State1>, otherStateMachineState :OtherState ->
+                    ...
+                }
+            )
+        onEnterStartStateMachine(...) // You can have multiple onEnterStartStateMachine
+        onActionStartStateMachine(...) // You can have multiple onActionStartStateMachine
+
+        // Custom conditions
+        condition({state.someString == "Hello" }){
+            // Everything inside this block only executes if the surounding condition is met 
+            // and the state machine is in the state as specified by the top level inState<State1>.
+            // 
+            //  you can have multiple of the dsl blocks, i.e. multiple on<Action> blocks and so on.
+            on<Action3>{ action, state -> ... } // you can have multiple on<Action>
+            onEnter{ state -> ... }
+            collectWhileInState(flow){ valueEmitedFromFlow, state -> ... } 
+            onActionEffect{ action, state -> ...}
+            onEnterEffect{ state -> ... }
+            collectWhileInStateEffect(flow){ valueEmitedFromFlow, state -> ... } 
+            onEnterStartStateMachine(...)
+            onActionStartStateMachine(...)
+
+            untilIdentityChanged({ state.id }) {
```

Comment:
Should this be moved out of `condition` and then we only have a "stub" here?

Reply:
Not sure if I understand what you mean ...

Reply:
Structure wise you have something like
```
inState {
  on // with description
  onEnter // with description
  collectWhileInState // with description
  ...

  conditition {

    on // NO description
    onEnter // NO description
    collectWhileInState // NO description

    untilIdentityChanged // with description
  }
}
```
My suggestion would be to move `untilIdentityChanged` one level higher and don't have a description for it inside condition

```
inState {
  on // with description
  onEnter // with description
  collectWhileInState // with description
  ...

  conditition {

    on // NO description
    onEnter // NO description
    collectWhileInState // NO description

    untilIdentityChanged // version of untilIdentityChanged that is only ran if the condition block is active, see below for more
  }
  
  untilIdentityChanged // with description
}
```